networkx/algorithms/summarization.py

```diff
@@ -210,3 +210,456 @@ def dedensify(G, threshold, prefix=None, copy=True):
+
+    Returns
+    -------
+    tuple of data structure representations: (list, list, dict, dict)
+
+    Notes
+    -----
+    Finds the AR-compatible grouping of graph `G` and returns data structures
+    to efficiently process those groupings.
+
+    An Attribute-edge compatible (AR-compatible) grouping is where
+    all nodes in each group have the same exact node attribute values and
+    have the same exact edge types to nodes in the same exact
+    groups.
+    """
+    edge_types = {
+        edge: tuple(attrs.get(attr) for attr in edge_attributes)
+        for edge, attrs in G.edges.items()
+    }
+    if not G.is_directed():
+        if G.is_multigraph():
+            edge_types.update(
+                ((e[1], e[0], e[2]), edge_types[e]) for e in list(edge_types)
+            )
+        else:
+            edge_types.update(
+                ((v, u), edge_types[(u, v)]) for (u, v) in list(edge_types)
+            )
+
+    group_lookup = {
+        node: tuple(attrs[attr] for attr in node_attributes)
+        for node, attrs in G.nodes.items()
+    }
+    groups = defaultdict(set)
+    for node, node_type in group_lookup.items():
+        groups[node_type].add(node)
+
+    neighbor_groups = snap_calculate_neighbor_groups(
+        G,
+        groups,
+        edge_types,
+        group_lookup,
+    )
+    eligible_groups = [
+        g_id for g_id in groups if snap_is_eligible(groups, neighbor_groups, g_id)
+    ]
+    while eligible_groups:
+        group_id = eligible_groups.pop()
+        groups = snap_split(
+            G, groups, neighbor_groups, edge_types, group_lookup, group_id
+        )
+        neighbor_groups = snap_calculate_neighbor_groups(
+            G,
+            groups,
+            edge_types,
+            group_lookup,
+        )
+        eligible_groups = [
+            g_id for g_id in groups if snap_is_eligible(groups, neighbor_groups, g_id)
+        ]
+    return (groups, neighbor_groups, edge_types)
+
+
+def snap_is_eligible(groups, neighbor_groups, group_id):
+    """
+    Determines if a group is eligible to be split.
+
+    A group is eligible to be split if all nodes in the group have edges of the same type(s)
+    with the same other groups.
+
+    Parameters
+    ----------
+    groups: dict
+        A dictionary of unique group IDs and their corresponding node groups
+    neighbor_groups: dict
+        A data structure indicating the number of edges a node has with the
+        groups in the current summarization of each edge type
+    group_id: object
+        ID of group to check for split eligibility
+
+    Returns
+    -------
+    boolean: Indicates if the group can/should be split into multiple groups
+    """
+    current_group = groups[group_id]
+    group_size = len(current_group)
+    for other_group_id in groups:
+        edge_counts = Counter()
+        for node in current_group:
+            edge_counts.update(neighbor_groups[node][other_group_id].keys())
```

Comment:
```suggestion
    node_neighbor_groups = neighbor_groups[node]
    current_group = groups[group_id]
    group_size = len(current_group)
    for other_group_id in groups:
        edge_counts = Counter()
        for node in current_group:
            edge_counts.update(node_neighbor_groups[other_group_id])
```
Feel free to rename as you like... this just moves the dict lookup outside the loop and removes the superfluous `keys`.